File: fusion/rules.py
# -*- coding: utf-8 -*-
"""Island analysis and the deterministic arms.

`classify`, `window_parts`, `RULES` and the per-arm `build` logic are ported
from eval/controlled_eval/chooser/reconstruct.py; `apply_guard` verbatim from
eval/controlled_eval/chooser/make_fixtures.py.

The one structural change from the research code is PROVENANCE, not logic: every
token this module produces is carried as a `(token, system_index, column_index)`
triple instead of a bare string, so `fuse.py` can name the raw provider word
behind each output token. Strip the triples with `toks()` and the sequences are
byte-identical to the research code's lists — which is what the conformance
suite proves.

Guard definition (frozen, spec §4.1): an island is guarded iff (a) after
removing every token in {ναι, δεν} the three candidate spans are identical,
(b) at least one span contains such a token, and (c) the arm's island output
contains none. Then the output is replaced by the span with the most critical
tokens (ties: the longest, then scribe > soniox > ours).
"""
from __future__ import annotations

import hashlib
import json

from . import categories as C
from .islands import column_class, islands, split_merge_columns
from .msa import vote_column

CRIT = ("ναι", "δεν")

# Rule names are the frozen Greek strings used in policy.json.
RULE_VOTE = "ψήφος"
RULE_R12 = "R1+R2"
RULE_SCRIBE = "Scribe"
RULE_DELETE = "σβήσε"


def toks(triples):
    """The bare token list behind a list of (token, system, column) triples."""
    return [t for t, _s, _c in triples]


def classify(cols, span, sm, ctx):
    """Port of reconstruct.classify — note it does NOT append |ATOMIC/|COMPOSITE
    to the TWO_/SOLO_/PAIRE_ signatures, unlike cats.py's variant."""
    spans = [tuple(x for x in (cols[i][k] for i in span) if x is not None)
             for k in range(3)]
    ragged = len({len(x) for x in spans}) > 1
    composite = ragged or any(i in sm for i in span)
    if len({C.norm_num(x) for x in spans}) == 1:
        return "NUM", spans
    if len({C.strip_fill(x) for x in spans}) == 1:
        return "FILL", spans
    if len({C.strip_dup(x, ctx) for x in spans}) == 1:
        return "DUP", spans
    if len({C.strip_dup(C.strip_fill(x), ctx) for x in spans}) == 1:
        return "LOW", spans
    sig = C.signature(spans)
    if sig.startswith(("TWO_", "SOLO_", "PAIRE_")):
        return sig, spans
    return sig + ("|COMPOSITE" if composite else "|ATOMIC"), spans


def _vote_source(col, token):
    """Which system supplied a voted token. Frozen priority scribe>soniox>ours."""
    for k in range(3):
        if col[k] == token:
            return k
    return None


def window_parts(cols, pivot: int):
    """(islands, tail, wsel, wsrc) — enough to rebuild any arm's text.

    `cols` is the aligned column list, `pivot` the consensus pivot index.
    Each island dict carries both the frozen plain-token fields the research
    code produced (`cat`, `spans`, `vote`, `r12`, `L`, `R`, `before`) and the
    provenance triples (`vote_tr`, `r12_tr`, `span_tr`, `before_tr`).
    """
    cols = [tuple(c) for c in cols]
    wsel = [vote_column(c, pivot)[0] for c in cols]
    wsrc = [(_vote_source(cols[i], wsel[i]) if wsel[i] is not None else None)
            for i in range(len(cols))]
    sm = split_merge_columns(cols)
    out, prev = [], 0
    for (s, e) in islands(cols):
        span = list(range(s, e))
        L = [cols[i][0] for i in range(max(0, s - 12), s)
             if column_class(cols[i]) == "agree"]
        R = [cols[i][0] for i in range(e, min(len(cols), e + 12))
             if column_class(cols[i]) == "agree"]
        cat, spans = classify(cols, span, sm, set(L[-3:]) | set(R[:3]))

        vote_tr = [(wsel[i], wsrc[i], i) for i in span if wsel[i] is not None]
        r12_tr = []
        for i in span:
            k = column_class(cols[i])
            if k == "unresolved_two":
                continue
            if k == "unresolved_three":
                t, src = cols[i][0], 0
            else:
                t, src = wsel[i], wsrc[i]
            if t is not None:
                r12_tr.append((t, src, i))
        span_tr = [[(cols[i][k], k, i) for i in span if cols[i][k] is not None]
                   for k in range(3)]
        before_tr = [(cols[i][0], 0, i) for i in range(prev, s)
                     if column_class(cols[i]) == "agree"]

        out.append({
            "s": s, "e": e, "cat": cat, "spans": [list(x) for x in spans],
            "vote": toks(vote_tr), "r12": toks(r12_tr),
            "L": " ".join(L[-8:]), "R": " ".join(R[:8]),
            "before": toks(before_tr),
            "vote_tr": vote_tr, "r12_tr": r12_tr, "span_tr": span_tr,
            "before_tr": before_tr,
        })
        prev = e
    tail_tr = [(cols[i][0], 0, i) for i in range(prev, len(cols))
               if column_class(cols[i]) == "agree"]
    return out, tail_tr, wsel, wsrc


RULES = {RULE_VOTE: lambda i: i["vote_tr"],
         RULE_R12: lambda i: i["r12_tr"],
         RULE_SCRIBE: lambda i: i["span_tr"][0],
         RULE_DELETE: lambda i: []}


def apply_guard(spans, out):
    """Verbatim from make_fixtures.apply_guard, over plain token lists."""
    core = [[t for t in sp if t not in CRIT] for sp in spans]
    if not (core[0] == core[1] == core[2]):
        return out, False
    n_crit = [sum(t in CRIT for t in sp) for sp in spans]
    if not any(n_crit):
        return out, False
    if any(t in CRIT for t in out):
        return out, False
    k = max(range(3), key=lambda i: (n_crit[i], len(spans[i]), -i))
    return list(spans[k]), True


def apply_guard_tr(isl, out_tr):
    """`apply_guard` on triples: the replacement span is a whole system's span,
    so its provenance is `span_tr[k]`."""
    spans = [list(x) for x in isl["spans"]]
    core = [[t for t in sp if t not in CRIT] for sp in spans]
    if not (core[0] == core[1] == core[2]):
        return out_tr, False
    n_crit = [sum(t in CRIT for t in sp) for sp in spans]
    if not any(n_crit):
        return out_tr, False
    if any(t in CRIT for t in toks(out_tr)):
        return out_tr, False
    k = max(range(3), key=lambda i: (n_crit[i], len(spans[i]), -i))
    return list(isl["span_tr"][k]), True


def row_sha(i):
    """The render-row hash the policy fixtures key on. Verbatim from
    make_fixtures.row_sha."""
    row = {"L": i["L"], "R": i["R"], "spans": i["spans"], "r12": i["r12"]}
    return hashlib.sha256(
        json.dumps(row, ensure_ascii=False, sort_keys=True).encode()).hexdigest()[:16]


def island_route(policy, cat):
    """(route, rule) for a category under a policy dict. Matches make_fixtures."""
    p = policy.get(cat)
    if not p:
        return "vote", None
    return p["mode"], (p.get("rule") if p["mode"] == "rule" else None)

File: src/lib/fusion/engine/rules.ts
/**
 * Island analysis and the deterministic arms, ported from `fusion/rules.py`.
 *
 * Every token here is carried as a `[token, system, column]` triple rather than
 * a bare string, so the driver can name the raw provider word behind each
 * output token. Strip the triples with `toks()` and the sequences are identical
 * to the research code's lists, which is what the conformance suite proves.
 *
 * Guard definition (frozen): an island is guarded iff, after removing every
 * token in {ναι, δεν}, the three candidate spans are identical, at least one
 * span contains such a token, and the arm's output contains none. The output is
 * then replaced by the span with the most critical tokens, ties going to the
 * longest, then to scribe over soniox over ours.
 */
import crypto from "crypto";
import * as C from "./categories.js";
import { columnClass, islands, splitMergeColumns } from "./islands.js";
import { voteColumn, type Column } from "./msa.js";
import { pyJsonDumps } from "./pyjson.js";

export const CRIT: readonly string[] = ["ναι", "δεν"];

/** Rule names are the frozen Greek strings used in `policy.json`. */
export const RULE_VOTE = "ψήφος";
export const RULE_R12 = "R1+R2";
export const RULE_SCRIBE = "Scribe";
export const RULE_DELETE = "σβήσε";

/** `[token, systemIndex, columnIndex]`. */
export type Triple = [string, number, number];

export function toks(triples: readonly Triple[]): string[] {
    return triples.map((t) => t[0]);
}

const key = (xs: readonly string[]) => JSON.stringify(xs);

/**
 * Port of the research `classify`. Note it does NOT append |ATOMIC or
 * |COMPOSITE to the TWO_, SOLO_ and PAIRE_ signatures, unlike the variant in
 * the category module.
 */
export function classify(
    cols: readonly Column[],
    span: readonly number[],
    sm: ReadonlySet<number>,
    ctx: ReadonlySet<string>,
): { cat: string; spans: string[][] } {
    const spans: string[][] = [];
    for (let k = 0; k < 3; k++) {
        const s: string[] = [];
        for (const i of span) {
            const e = cols[i][k];
            if (e !== null) s.push(e);
        }
        spans.push(s);
    }

    const ragged = new Set(spans.map((x) => x.length)).size > 1;
    const composite = ragged || span.some((i) => sm.has(i));

    if (new Set(spans.map((x) => key(C.normNum(x)))).size === 1) return { cat: "NUM", spans };
    if (new Set(spans.map((x) => key(C.stripFill(x)))).size === 1) return { cat: "FILL", spans };
    if (new Set(spans.map((x) => key(C.stripDup(x, ctx)))).size === 1) return { cat: "DUP", spans };
    if (new Set(spans.map((x) => key(C.stripDup(C.stripFill(x), ctx)))).size === 1) {
        return { cat: "LOW", spans };
    }

    const sig = C.signature(spans);
    if (sig.startsWith("TWO_") || sig.startsWith("SOLO_") || sig.startsWith("PAIRE_")) {
        return { cat: sig, spans };
    }
    return { cat: sig + (composite ? "|COMPOSITE" : "|ATOMIC"), spans };
}

/** Which system supplied a voted token. Frozen priority scribe > soniox > ours. */
function voteSource(col: Column, token: string | null): number | null {
    for (let k = 0; k < 3; k++) {
        if (col[k] === token) return k;
    }
    return null;
}

export interface Island {
    s: number;
    e: number;
    cat: string;
    spans: string[][];
    vote: string[];
    r12: string[];
    L: string;
    R: string;
    before: string[];
    vote_tr: Triple[];
    r12_tr: Triple[];
    span_tr: Triple[][];
    before_tr: Triple[];
}

export interface WindowParts {
    islands: Island[];
    tail_tr: Triple[];
    wsel: (string | null)[];
    wsrc: (number | null)[];
}

/** Islands, tail, and the voted selection: enough to rebuild any arm's text. */
export function windowParts(cols: readonly Column[], pivot: number): WindowParts {
    const wsel = cols.map((c) => voteColumn(c, pivot).token);
    const wsrc = cols.map((c, i) => (wsel[i] !== null ? voteSource(c, wsel[i]) : null));
    const sm = splitMergeColumns(cols);

    const out: Island[] = [];
    let prev = 0;
    for (const [s, e] of islands(cols)) {
        const span: number[] = [];
        for (let i = s; i < e; i++) span.push(i);

        const L: string[] = [];
        for (let i = Math.max(0, s - 12); i < s; i++) {
            if (columnClass(cols[i]) === "agree") L.push(cols[i][0] as string);
        }
        const R: string[] = [];
        for (let i = e; i < Math.min(cols.length, e + 12); i++) {
            if (columnClass(cols[i]) === "agree") R.push(cols[i][0] as string);
        }

        const ctx = new Set<string>([...L.slice(-3), ...R.slice(0, 3)]);
        const { cat, spans } = classify(cols, span, sm, ctx);

        const vote_tr: Triple[] = [];
        for (const i of span) {
            if (wsel[i] !== null) vote_tr.push([wsel[i] as string, wsrc[i] as number, i]);
        }

        const r12_tr: Triple[] = [];
        for (const i of span) {
            const k = columnClass(cols[i]);
            if (k === "unresolved_two") continue;
            let t: string | null;
            let src: number | null;
            if (k === "unresolved_three") {
                t = cols[i][0];
                src = 0;
            } else {
                t = wsel[i];
                src = wsrc[i];
            }
            if (t !== null) r12_tr.push([t, src as number, i]);
        }

        const span_tr: Triple[][] = [];
        for (let k = 0; k < 3; k++) {
            const row: Triple[] = [];
            for (const i of span) {
                const e2 = cols[i][k];
                if (e2 !== null) row.push([e2, k, i]);
            }
            span_tr.push(row);
        }

        const before_tr: Triple[] = [];
        for (let i = prev; i < s; i++) {
            if (columnClass(cols[i]) === "agree") before_tr.push([cols[i][0] as string, 0, i]);
        }

        out.push({
            s, e, cat, spans,
            vote: toks(vote_tr),
            r12: toks(r12_tr),
            L: L.slice(-8).join(" "),
            R: R.slice(0, 8).join(" "),
            before: toks(before_tr),
            vote_tr, r12_tr, span_tr, before_tr,
        });
        prev = e;
    }

    const tail_tr: Triple[] = [];
    for (let i = prev; i < cols.length; i++) {
        if (columnClass(cols[i]) === "agree") tail_tr.push([cols[i][0] as string, 0, i]);
    }
    return { islands: out, tail_tr, wsel, wsrc };
}

export const RULES: Record<string, (i: Island) => Triple[]> = {
    [RULE_VOTE]: (i) => i.vote_tr,
    [RULE_R12]: (i) => i.r12_tr,
    [RULE_SCRIBE]: (i) => i.span_tr[0],
    [RULE_DELETE]: () => [],
};

/**
 * Python's `max(range(3), key=...)` keeps the FIRST index on a tie and compares
 * the key tuples lexicographically, so the order of these three comparisons is
 * the tie-break: most critical tokens, then longest span, then the lower system
 * index. Written out because a JavaScript sort would neither be stable in the
 * same way nor compare tuples at all.
 */
function bestGuardSpan(nCrit: readonly number[], spans: readonly (readonly string[])[]): number {
    let best = 0;
    for (let i = 1; i < 3; i++) {
        const a: [number, number, number] = [nCrit[i], spans[i].length, -i];
        const b: [number, number, number] = [nCrit[best], spans[best].length, -best];
        if (a[0] > b[0] || (a[0] === b[0] && (a[1] > b[1] || (a[1] === b[1] && a[2] > b[2])))) {
            best = i;
        }
    }
    return best;
}

function guardable(
    spans: readonly (readonly string[])[],
    outTokens: readonly string[],
): number | null {
    const core = spans.map((sp) => sp.filter((t) => !CRIT.includes(t)));
    if (key(core[0]) !== key(core[1]) || key(core[1]) !== key(core[2])) return null;
    const nCrit = spans.map((sp) => sp.reduce((n, t) => n + (CRIT.includes(t) ? 1 : 0), 0));
    if (!nCrit.some((n) => n > 0)) return null;
    if (outTokens.some((t) => CRIT.includes(t))) return null;
    return bestGuardSpan(nCrit, spans);
}

/** The guard over plain token lists. */
export function applyGuard(
    spans: readonly (readonly string[])[],
    out: readonly string[],
): { out: string[]; fired: boolean } {
    const k = guardable(spans, out);
    if (k === null) return { out: [...out], fired: false };
    return { out: [...spans[k]], fired: true };
}

/**
 * The guard on triples. The replacement span is one whole system's span, so its
 * provenance is that system's `span_tr`.
 */
export function applyGuardTr(
    isl: Island,
    outTr: readonly Triple[],
): { out: Triple[]; fired: boolean } {
    const k = guardable(isl.spans, toks(outTr));
    if (k === null) return { out: [...outTr], fired: false };
    return { out: [...isl.span_tr[k]], fired: true };
}

/** The render-row hash the policy fixtures key on. */
export function rowSha(i: Island): string {
    const row = { L: i.L, R: i.R, spans: i.spans, r12: i.r12 };
    return crypto.createHash("sha256")
        .update(pyJsonDumps(row), "utf8")
        .digest("hex")
        .slice(0, 16);
}

/** `[route, rule]` for a category under a policy dict. */
export function islandRoute(
    policy: Record<string, { mode: string; rule?: string }>,
    cat: string,
): [string, string | null] {
    const p = policy[cat];
    if (!p) return ["vote", null];
    return [p.mode, p.mode === "rule" ? (p.rule ?? null) : null];
}
